paddle/fluid/platform/device_context.h

```diff
@@ -207,6 +214,20 @@ class CudnnWorkspaceHandle {
                          required_workspace_len);
   }
 
+  /*! \brief Thread which call RunFuncReleaseMem() would acquire the lock first
+   *  before invoking cudnn functions and release gpu memory after running
+   *  functions */
+  template <typename Callback>
+  inline void RunFuncReleaseMem(Callback&& cudnn_func,
+                                size_t required_workspace_len) {
+    if (!guard_) {
+      guard_.reset(new std::lock_guard<std::mutex>(holder_->Mutex()));
+    }
+    holder_->RunFuncImpl(std::forward<Callback>(cudnn_func),
+                         required_workspace_len);
+    holder_->ResetWorkspace();
```

Comment:
The execution of CUDA code is async, so you should call `cudaStreamSync` before reset workspace.

Reply:
Maybe we should call this method `RunFuncSync`? Because this function may be only used when cudnn exhaustive searching, and `cudnnFindConvolutionForwardAlgorithmEx` is a synchronized method as described in [here](https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnFindConvolutionForwardAlgorithmEx).

Reply:
@chengduoZH Done. I added inside ResetWorkspace() to enforce users to call cudaStreamSynchronize.

@sneaxiy Done